src/io/flutter/inspector/InspectorService.java

```diff
@@ -36,8 +57,127 @@
  * inspector code running in the IDE.
  */
 public class InspectorService implements Disposable {
+
+  public static class Location {
+
+    public Location(VirtualFile file, int line, int column, int offset) {
+      this.file = file;
+      this.line = line;
+      this.column = column;
+      this.offset = offset;
+    }
+
+    private final VirtualFile file;
+    public final int line;
+    public final int column;
+    private final int offset;
+
+    public int getLine() {
+      return line;
+    }
+
+    public int getColumn() {
+      return column;
+    }
+
+    public int getOffset() {
+      return offset;
+    }
+
+    public VirtualFile getFile() {
+      return file;
+    }
+
+    public String getPath() {
+      return toSourceLocationUri(file.getPath());
+    }
+
+    public XSourcePosition getXSourcePosition() {
+      if (file == null) {
+        return null;
+      }
+      final int line = getLine();
+      final int column = getColumn();
+      if (line < 0 || column < 0) {
+        return null;
+      }
+      return XSourcePositionImpl.create(file, line - 1, column - 1);
+    }
+
+    /**
+     * Returns a location for a FlutterOutline object that makes a best effort
+     * to be compatible with the locations generated by the flutter kernel
+     * transformer to track creation locations.
+     */
+    public static @Nullable
+    InspectorService.Location outlineToLocation(Editor editor, FlutterOutline outline) {
+      if (!(editor instanceof EditorEx)) return null;
+      EditorEx editorEx = (EditorEx)editor;
+      return outlineToLocation(editor.getProject(), editorEx.getVirtualFile(), outline, editor.getDocument());
+    }
+
+    public static InspectorService.Location outlineToLocation(Project project,
+                                                              VirtualFile file,
+                                                              FlutterOutline outline,
+                                                              Document document) {
+      if (file == null) return null;
+      if (document == null) return null;
+      if (outline == null || outline.getClassName() == null) return null;
+      final int documentLength = document.getTextLength();
+      int nodeOffset = Math.max(0, Math.min(outline.getCodeOffset(), documentLength));
+      final int nodeEndOffset = Math.max(0, Math.min(outline.getCodeOffset() + outline.getCodeLength(), documentLength));
+
+      // The DartOutline will give us the offset of
+      // 'child: Foo.bar(...)'
+      // but we need the offset of 'bar(...)' for consistentency with the
+      // Flutter kernel transformer.
+      if (outline.getClassName() != null) {
+        final PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
+
```

Comment:
`psiFile` can be null; best to guard against that.

Reply:
good point. done